parla/task_runtime.py

```diff
@@ -1255,23 +1254,54 @@ def run_mapper(self):
                 # move to the mapped task scheduling.
                 break
 
-    def run_deviceq_mapper(self):
+    def _schedule_tasks(self):
+        """ Currently this doesn't do any intelligent scheduling (ordering).
+            Dequeue all ready tasks and send them to device queues in order.
+        """
+        print("[Scheduler] Schedule Phase")
         while True:
             task: Optional[TaskBase] = self._dequeue_task()
             if not task or not task.assigned:
                 logger.debug("Task %r: Failed to assign", task)
                 break
             for d in task.req.devices:
-                print("[Scheduler] Task, ", task.name, ", is enqueued onto a device Q.", sep='')
-                worker = self._worker_threads[task.req.env_no]
-                worker._enqueue_task_local(task)
+                print(f"[Scheduler] Enqueuing {task.name} to device {d}")
+                self._device_queues[d].append(task)
+
+    def _launch_tasks(self):
+        """ Iterate through free devices and launch tasks on them
+        """
+        print("[Scheduler] Launch Phase")
+        with self._monitor:
+            for dev, queue in self._device_queues.items():
+                # Make sure there's an available WorkerThread
+                if len(self._free_worker_threads) == 0:
+                    break
+
+                # Hack to make launching work for now
+                # Only works for single-device tasks, and only one task at a time on each device
+                # TODO: Make this not terrible
+                if self._available_resources._occupancy_dict[dev] == False: # if not occupied
+                    if len(queue) > 0: # if there are tasks on the queue
+                        task = queue.pop() # grab a task
+                        worker = self._free_worker_threads.pop() # grab a worker
+                        print(f"[Scheduler] Launching {task.name} on WorkerThread {worker.index}")
+                        self._available_resources._occupancy_dict[dev] = True # mark the device as occupied
+                        worker.assign_task(task) # assign the task to the worker (this notifies the worker's monitor)
+                        print(f"[Scheduler] Launched {worker.task.name}")
 
     def run(self) -> None:
         # noinspection PyBroadException
         try: # Catch all exception to report them usefully
+            i = 0
             while self._should_run:
-                self.run_mapper()
-                self.run_deviceq_mapper()
+                self._map_tasks()
+                self._schedule_tasks()
+                self._launch_tasks()
+                print("[Scheduler] Sleeping!")
+                time.sleep(1) # TODO: Delete this. Why the fuck doesn't this release the GIL.
```

Comment:
we need to adjust this to sleep for a dynamically set `period`

Reply:
also, ignore my angry comment. It does release the GIL. I'm just dumb.

Reply:
If it works, please remove the comment. I guess we tried to avoid sleep, didn't we? (This is real question)
How about using queue length before calling any function as Will suggested?

Reply:
See https://github.com/ut-parla/Parla.py/pull/99#discussion_r801852437

I'm currently leaving it as a sleep until we determine we need a different mechanism (which we probably will, but we don't know that yet).